browser/brave_rewards/checkout_dialog.cc

```diff
@@ -0,0 +1,343 @@
+                      errors::kRewardsNotInitialized);
+    return;
+  }
+
+  const auto& display_items =
+      spec->GetDisplayItems(request_->state()->selected_app());
+  for (size_t i = 0; i < display_items.size(); i++) {
+    DCHECK((*display_items[i])->sku.has_value());
+    auto item = ledger::type::SKUOrderItem::New();
+
+    item->sku = (*display_items[i])->sku.value();
+    item->quantity = 1;
+    items_.push_back(std::move(item));
+  }
+
+  auto callback = base::BindOnce(&OnSKUProcessed, request_);
+
+  rewards_service->ProcessSKU(
+      std::move(items_), ledger::constant::kWalletUphold, std::move(callback));
+}
+
+void CheckoutDialogHandler::GetRewardsParameters(const base::ListValue* args) {
+  if (auto* service = GetRewardsService()) {
+    AllowJavascript();
+    service->GetRewardsParameters(
+        base::Bind(&CheckoutDialogHandler::OnGetRewardsParameters,
+                   weak_factory_.GetWeakPtr()));
+  }
+}
+
+void CheckoutDialogHandler::OnGetRewardsParameters(
+    ledger::type::RewardsParametersPtr parameters) {
+  if (!IsJavascriptAllowed()) {
+    return;
+  }
+
+  base::Value data(base::Value::Type::DICTIONARY);
+  if (parameters) {
+    data.SetDoubleKey("rate", parameters->rate);
+    data.SetDoubleKey("lastUpdated", base::Time::Now().ToJsTimeIgnoringNull());
+  }
+  FireWebUIListener("rewardsParametersUpdated", data);
+}
+
+void CheckoutDialogHandler::GetWalletBalance(const base::ListValue* args) {
+  if (auto* service = GetRewardsService()) {
+    AllowJavascript();
+    service->FetchBalance(
+        base::BindOnce(&CheckoutDialogHandler::OnFetchBalance,
+                       weak_factory_.GetWeakPtr()));
+  }
+}
+
+void CheckoutDialogHandler::OnFetchBalance(
+    const ledger::type::Result result,
+    ledger::type::BalancePtr balance) {
+  if (!IsJavascriptAllowed()) {
+    return;
+  }
+
+  base::Value data(base::Value::Type::DICTIONARY);
+  data.SetDoubleKey("total", balance->total);
+
+  FireWebUIListener("walletBalanceUpdated", data);
+}
+
+void CheckoutDialogHandler::GetExternalWallet(const base::ListValue* args) {
+  if (auto* service = GetRewardsService()) {
+    AllowJavascript();
+    service->GetExternalWallet(
+        base::BindOnce(&CheckoutDialogHandler::OnGetExternalWallet,
+                       weak_factory_.GetWeakPtr()));
+  }
+}
+
+void CheckoutDialogHandler::OnGetExternalWallet(
+    const ledger::type::Result result,
+    ledger::type::ExternalWalletPtr wallet) {
+  if (!IsJavascriptAllowed()) {
+    return;
+  }
+
+  base::Value data(base::Value::Type::DICTIONARY);
+
+  if (wallet) {
+    data.SetIntKey("status", static_cast<int>(wallet->status));
+  }
+  FireWebUIListener("externalWalletUpdated", data);
+}
+
+void ShowCheckoutDialog(base::WeakPtr<PaymentRequest> request) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  if (!request) {
+    return;
+  }
+
+  WebContents* initiator = request->web_contents();
```

Comment:
I don't deeply understand this concept, but chrome uses `initiator_web_contents->GetResponsibleWebContents()` https://source.chromium.org/chromium/chromium/src/+/master:components/constrained_window/constrained_window_views.cc;drc=43b334ad1641a94cb11dec0d8023be47b16a66c0;l=177